### PR TITLE
fixed data selection by array of indecies

### PR DESCRIPTION
**Pull request by tarurar**

### CODE REVIEW

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 1]
```diff
@@ -123,28 +122,38 @@ public static AzureMultiIndex Create(string partitionKey, string rowKey, params
 
 			return result;
 		}
-
-
 	}
 
 	public static class AzureIndexUtils
 	{
+	    public static Tuple<string, string> ToTuple(this IAzureIndex src)
```
Comment (MikhailGerasimov) [inline src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Doubt that this is a useful method for sharing

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 2]
```diff
@@ -156,41 +165,35 @@ public static Task<T> DeleteAsync<T>(this INoSQLTableStorage<T> tableStorage, IA
...
+		public static async Task<T> GetDataAsync<T>(this INoSQLTableStorage<T> tableStorage, IAzureIndex index) where T : class, ITableEntity, new()
 		{
 			if (index == null)
 				return null;
 
 			return await tableStorage.GetDataAsync(index.PrimaryPartitionKey, index.PrimaryRowKey);
 		}
 
-		public async static Task<T> GetDataAsync<T>(this INoSQLTableStorage<T> tableStorage, INoSQLTableStorage<AzureIndex> indexTableStorage,
+		public static async Task<T> GetDataAsync<T>(this INoSQLTableStorage<T> tableStorage, INoSQLTableStorage<AzureIndex> indexTableStorage,
```
Comment (MikhailGerasimov) [inline src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Delete async|await. Just return a task

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 3]
```diff
@@ -156,41 +165,35 @@ public static Task<T> DeleteAsync<T>(this INoSQLTableStorage<T> tableStorage, IA
...
-		public async static Task<T> ReplaceAsync<T>(this INoSQLTableStorage<AzureIndex> indexTableStorage, string indexPartitionKey, string indexRowKey, INoSQLTableStorage<T> tableStorage, Func<T, T> action) where T : class, ITableEntity, new()
+		public static async Task<T> ReplaceAsync<T>(this INoSQLTableStorage<AzureIndex> indexTableStorage, string indexPartitionKey, string indexRowKey, INoSQLTableStorage<T> tableStorage, Func<T, T> action) where T : class, ITableEntity, new()
 		{
 			var indexEntity = await indexTableStorage.GetDataAsync(indexPartitionKey, indexRowKey);
 			return await tableStorage.ReplaceAsync(indexEntity, action);
 		}
 
-
-		public async static Task<T> ReplaceAsync<T>(this INoSQLTableStorage<T> tableStorage, IAzureIndex index, Func<T, T> action) where T : class, ITableEntity, new()
+		public static async Task<T> ReplaceAsync<T>(this INoSQLTableStorage<T> tableStorage, IAzureIndex index, Func<T, T> action) where T : class, ITableEntity, new()
```
Comment (MikhailGerasimov) [inline src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Delete async|await

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 4]
```diff
@@ -123,28 +122,38 @@ public static AzureMultiIndex Create(string partitionKey, string rowKey, params
...
+        /// <param name="pieces"></param>
+        /// <param name="filter"></param>
+        /// <returns></returns>
 		public static async Task<IEnumerable<T>> GetDataAsync<T>(this INoSQLTableStorage<T> tableStorage,
 				 IEnumerable<IAzureIndex> indices, int pieces = 15, Func<T, bool> filter = null) where T : ITableEntity, new()
 		{
-			var idx = indices.ToArray();
-			if (idx.Length == 0)
-				return new T[0];
+            var idx = indices.ToArray();
```
Comment (MikhailGerasimov) [inline src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Use Any()  to check. ToArray() creates a copy of all values only for one element

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 5]
```diff
@@ -123,28 +122,38 @@ public static AzureMultiIndex Create(string partitionKey, string rowKey, params
...
+            if (idx.Length == 0)
+                return new T[0];
```
Comment (MikhailGerasimov) [inline src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Return Array.Empty<>

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 6]
```diff
@@ -123,28 +122,38 @@ public static AzureMultiIndex Create(string partitionKey, string rowKey, params
...
 
-			var partitionKey = idx.First().PrimaryPartitionKey;
-			var rowKeys = idx.Select(itm => itm.PrimaryRowKey).ToArray();
-			return await tableStorage.GetDataAsync(partitionKey, rowKeys, pieces, filter);
-		}
+            return await tableStorage.GetDataAsync(idx.Select(x => x.ToTuple()), pieces, filter);
```
Comment (MikhailGerasimov) [inline src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Delete async | await

File: src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs [hunk 7]
```diff
@@ -134,17 +133,23 @@ public static Task<T> DeleteAsync<T>(this INoSQLTableStorage<T> tableStorage, IA
...
-			if (idx.Length == 0)
-				return new T[0];
+            if (!indices.Any())
+                return Task.FromResult(Enumerable.Empty<T>());
 
-			var partitionKey = idx.First().PrimaryPartitionKey;
-			var rowKeys = idx.Select(itm => itm.PrimaryRowKey).ToArray();
-			return await tableStorage.GetDataAsync(partitionKey, rowKeys, pieces, filter);
-		}
+            return tableStorage.GetDataAsync(indices.Select(x => x.ToTuple()), pieces, filter);
```
[the remark this answers is not in the document]
Comment (KonstantinRyazantsev) [reply src/Lykke.AzureStorage/Tables/Templates/Index/AzureIndex.cs]: Morevoer ```Any()``` check could be removed easily. GetDataAsync will do almost nothing if empty ```kyes``` is passed.

**Pull request merged**